flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+
+import numpy as np
+import torch
+from pytorch_lightning.trainer.states import RunningStage
+from pytorch_lightning.utilities.enums import LightningEnum
+from torch.nn import Module
+from torchvision.datasets.folder import has_file_allowed_extension, make_dataset
+
+from flash.data.auto_dataset import AutoDataset, BaseAutoDataset, IterableAutoDataset
+from flash.data.properties import ProcessState, Properties
+from flash.data.utils import CurrentRunningStageFuncContext
+
+
+def has_len(data: Union[Sequence[Any], Iterable[Any]]) -> bool:
+    try:
+        len(data)
+        return True
+    except (TypeError, NotImplementedError):
+        return False
+
+
+@dataclass(unsafe_hash=True, frozen=True)
+class LabelsState(ProcessState):
+
+    labels: Optional[Sequence[str]]
+
+
+class MockDataset:
+
+    def __init__(self):
+        self.metadata = {}
+
+    def __setattr__(self, key, value):
+        if key != 'metadata':
+            self.metadata[key] = value
+        else:
+            object.__setattr__(self, key, value)
+
+
+DATA_TYPE = TypeVar("DATA_TYPE")
+
+
+class DataSource(Generic[DATA_TYPE], Properties, Module):
+
+    def load_data(self,
+                  data: DATA_TYPE,
+                  dataset: Optional[Any] = None) -> Union[Sequence[Mapping[str, Any]], Iterable[Mapping[str, Any]]]:
+        """Loads entire data from Dataset. The input ``data`` can be anything, but you need to return a Mapping.
+
+        Example::
+
+            # data: "."
+            # output: [("./cat/1.png", 1), ..., ("./dog/10.png", 0)]
+
+            output: Mapping = load_data(data)
+
+        """
+        return data
+
+    def load_sample(self, sample: Mapping[str, Any], dataset: Optional[Any] = None) -> Any:
+        """Loads single sample from dataset"""
+        return sample
+
+    def to_datasets(
+        self,
+        train_data: Optional[DATA_TYPE] = None,
+        val_data: Optional[DATA_TYPE] = None,
+        test_data: Optional[DATA_TYPE] = None,
+        predict_data: Optional[DATA_TYPE] = None,
+    ) -> Tuple[Optional[BaseAutoDataset], ...]:
+        train_dataset = self.generate_dataset(train_data, RunningStage.TRAINING)
+        val_dataset = self.generate_dataset(val_data, RunningStage.VALIDATING)
+        test_dataset = self.generate_dataset(test_data, RunningStage.TESTING)
+        predict_dataset = self.generate_dataset(predict_data, RunningStage.PREDICTING)
+        return train_dataset, val_dataset, test_dataset, predict_dataset
+
+    def generate_dataset(
+        self,
+        data: Optional[DATA_TYPE],
+        running_stage: RunningStage,
+    ) -> Optional[Union[AutoDataset, IterableAutoDataset]]:
+        is_none = data is None
+
+        if isinstance(data, Sequence):
+            is_none = data[0] is None
+
+        if not is_none:
+            from flash.data.data_pipeline import DataPipeline
+
+            mock_dataset = MockDataset()
```

Comment:
Need a mock dataset because the actual dataset object doesn't exist yet. Later on this can probably be refactored with some thought to just use the state system or something like that, but fo now I have used typing.cast :smiley: